pkg/services/tree/signature.go

```diff
@@ -26,15 +36,51 @@ func (s *Service) verifyClient(req interface{}, cid *cidSDK.ID, rawKey []byte) e
 		return fmt.Errorf("can't get container %s: %w", cid, err)
 	}
 
-	pub, err := keys.NewPublicKeyFromBytes(rawKey, elliptic.P256())
-	if err != nil {
-		return fmt.Errorf("invalid public key: %w", err)
-	}
+	if len(rawBearer) == 0 { // must be signed by the owner
+		// No error is expected because `VerifyDataWithSource` checks the signature.
+		// However, we may use different algorithms in the future, thus this check.
+		pub, err := keys.NewPublicKeyFromBytes(req.GetSignature().GetKey(), elliptic.P256())
+		if err != nil {
+			return fmt.Errorf("invalid public key: %w", err)
+		}
 
-	actualID := ownerSDK.NewIDFromPublicKey((*ecdsa.PublicKey)(pub))
-	if !actualID.Equal(cnr.OwnerID()) {
-		return errors.New("`Move` request must be signed by a container owner")
+		actualID := ownerSDK.NewIDFromPublicKey((*ecdsa.PublicKey)(pub))
+		if !actualID.Equal(cnr.OwnerID()) {
+			return errors.New("request must be signed by a container owner")
+		}
+		return nil
 	}
 
+	var bearer bearerSDK.BearerToken
+	if err := bearer.Unmarshal(rawBearer); err != nil {
+		return fmt.Errorf("invalid bearer token: %w", err)
+	}
+	if !bearer.OwnerID().Equal(cnr.OwnerID()) {
+		return errors.New("bearer token must be signed by the container owner")
+	}
+	if bCid := bearer.EACLTable().CID(); bCid != nil && !bCid.Equal(cid) {
+		return errors.New("bearer token is created for another container")
+	}
+	if err := bearer.VerifySignature(); err != nil {
+		return fmt.Errorf("invalid bearer token signature: %w", err)
+	}
+	action := eacl.NewValidator().CalculateAction(new(eacl.ValidationUnit).
+		WithEACLTable(bearer.EACLTable()).
+		WithContainerID(cid).
+		WithRole(eacl.RoleOthers).
```

Comment:
why the role is `RoleOthers` and is fixed?

Reply:
The last rule in the bearer EACL should be `deny all`, otherwise `CalculateAction` returns true.
First way is to use `RoleOthers` for this purpose/
Second way is to extend `CalculateAction` with the default value or use a second boolean return signifying that `CalculateAction` returned from some rule and not by default. I think the second variant is worth implementing, thoughts? @cthulhu-rider @alexvanin

Reply:
2nd option LGTM.

Reply:
@fyrchik lets link with https://github.com/nspcc-dev/neofs-sdk-go/issues/243?

Reply:
Added